File: divide_conquer/01_dc_longest_common_prefix/dc_longest_common_prefix.py
import sys
from typing import List


class Solution:
    def longest_common_prefix(self, strs: List[str]) -> str:
        if not self._check_constraints(strs):
            raise ValueError('Constraints violated')
        min_len = sys.maxsize
        for i in range(len(strs)):
            min_len = min(len(strs[i]), min_len)
        low, high = 1, len(strs)
        while low <= high:
            mid = (low + high) // 2
            if self._is_common_prefix(strs, mid):
                low = mid + 1
            else:
                high = mid - 1
        return strs[0][:(low + high) // 2]

    def _check_constraints(self, strs: List[str]) -> bool:
        if len(strs) < 1 or len(strs) > 200:
            return False
        for i in range(len(strs)):
            if len(strs[i]) < 0 or len(strs[i]) > 200:
                return False
            for ch in strs[i]:
                if ch < 'A' or (ch > 'Z' and ch < 'a') or ch > 'z':
                    return False
        return True

    def _is_common_prefix(self, strs: List[str], len: int) -> bool:
        min_str = strs[0][:len]
        if not all(s.startswith(min_str) for s in strs[1:]):
            return False
        return True

File: divide_conquer/01_dc_longest_common_prefix/dc_longest_common_prefix.cpp
#include <iostream>
#include <stdexcept>
#include <cmath>
#include <climits>
#include <vector>
#include <algorithm>
#include <string>

using namespace std;


class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        if (!checkConstraints(strs)) {
            throw std::runtime_error("Constraints violated");
        }
        return lcp(strs, 0, strs.size() - 1);
    }

private:
    bool checkConstraints(const vector<string>& strs) {
        if (strs.size() < 1 || strs.size() > 200) return false;
        for (int i = 0; i < strs.size(); i++) {
            if (strs[i].size() < 0 || strs[i].size() > 200) return false;
            for (auto& ch : strs[i]) {
                if (ch < 'A' || (ch > 'Z' && ch < 'a') || ch > 'z') return false;
            }
        }
        return true;
    }
    string lcp(vector<string>& strs, int left, int right) {
        if (left == right) {
            // conquer
            return strs[left];
        } else {
            // divide
            int mid = (left + right) / 2;
            string lcpLeft = lcp(strs, left, mid);
            string lcpRight = lcp(strs, mid + 1, right);
            return commonPrefix(lcpLeft, lcpRight);
        }
    }
    string commonPrefix(const string& left, const string& right) {
        auto pair = mismatch(left.begin(), left.end(), right.begin(), right.end());
        int minLen = pair.first - left.begin();
        return left.substr(0, minLen);
    }
};

// int main(int argc, char **argv) {
//     Solution* solution;
//     return 0;
// }
